<div align="center">
    <img src="https://i.pinimg.com/736x/04/41/52/0441528e0396b38e72d182acc0529f17.jpg" width="300px" alt="Logo University of Pisa">
</div>



## ❗**Disclaimer**❗

**Important:** This project was originally designed to handle a 2GB dataset. I have implemented extensive memory optimization techniques—including the use of `swifter` library, data chunking, and various code optimizations—to achieve this goal. Despite these efforts, processing the full 2GB dataset is currently not feasible in the available environment (Google Colab), as the **Preprocessing** step eventually exhausts the available RAM. Therefore, this demonstration uses a smaller dummy data to showcase the system's functionality.

**Zahra Omrani**

## Introduction

This notebook, titled "Tintinify", is an information retrieval system. The name was inspired by the curious and adventurous spirit of Tintin.This project represents the culmination of my learnings during the IRCV course at UNIPI, taught by Professor Nicola Tonellotto during my Master’s studies. Tintinify is engineered for efficient exploration and indexing of large-scale datasets.

### Key Features:
- **Document Preprocessing:** Implements robust methods for cleaning, normalizing, and tokenizing documents to prepare them for indexing.
- **Advanced Indexing Structures:** Constructs efficient data structures such as an inverted index, direct index, and lexicon to enable fast and accurate document retrieval.
- **BM25 Scoring:** Utilizes the BM25 algorithm to rank documents based on their relevance to a query.
- **DAAT Algorithm:** Applies the Document-At-A-Time (DAAT) approach for efficient query processing across large datasets.



## Dataset Overview

For my Information Retrieval project, I am using the **MS MARCO Passage Ranking dataset** from the **TREC 2020 Deep Learning Track**. This dataset is widely used for training and evaluating deep learning-based retrieval models. It consists of passages extracted from web documents and is designed for passage ranking tasks, where given a query, the goal is to retrieve the most relevant passages. The dataset provides a large-scale training set with human-labeled relevance judgments and is formatted as a **TSV file** containing passage IDs and their corresponding text. The MS MARCO dataset is frequently used in research for improving retrieval models and evaluating ranking methods.

## Dataset Properties

| Description  | Filename          | File Size | Num Records | Format               |
|-------------|------------------|-----------|-------------|----------------------|
| Collection  | collection.tar.gz | 2.9 GB    | 8,841,823   | TSV: pid, passage   |

The dataset is structured to facilitate large-scale training for passage retrieval tasks, making it a valuable resource for developing ranking models in Information Retrieval.


## Dataset Manager

In this cell we download and extracts a dataset directly into Google Drive using Google Colab.

- Mounts Google Drive.
- Checks if the dataset is already downloaded and extracted.
- Downloads the dataset if missing.
- Extracts the dataset for permanent storage.

In [1]:
import os
import tarfile
import requests
from google.colab import drive

def download_and_extract_to_drive(url, drive_folder, filename):
    """
    Downloads and extracts a dataset directly to Google Drive.

    Parameters:
        url (str): URL of the dataset.
        drive_folder (str): Google Drive path where the dataset will be saved.
        filename (str): Name of the dataset file.

    Behavior:
        - Mounts Google Drive.
        - Checks if the dataset is already downloaded and extracted in Drive.
        - Downloads it to Drive if missing.
        - Extracts it inside Drive for permanent storage.
    """
    # Mount Google Drive
    drive.mount('/content/drive')

    # Define paths in Google Drive
    drive_path = f"/content/drive/MyDrive/{drive_folder}"
    file_path = os.path.join(drive_path, filename)
    extract_path = os.path.join(drive_path, filename.replace('.tar.gz', ''))  # Extracted folder

    # Create the directory if it doesn’t exist
    os.makedirs(drive_path, exist_ok=True)

    # Download if the file does not exist
    if not os.path.exists(file_path):
        print(f"Downloading {filename} to Google Drive...")
        response = requests.get(url, stream=True)
        response.raise_for_status()
        with open(file_path, 'wb') as f:
            for chunk in response.iter_content(chunk_size=8192):
                f.write(chunk)
        print("Download completed.")

    else:
        print(f"{filename} already exists in Google Drive.")

    # Extract if not already extracted
    if not os.path.exists(extract_path):
        print(f"Extracting {filename} in Google Drive...")
        with tarfile.open(file_path, 'r:gz') as tar:
            tar.extractall(path=extract_path)
        print("Extraction completed.")
    else:
        print(f"{filename} is already extracted in Google Drive.")

    print(f"Dataset is ready at: {extract_path}")

# Dataset parameters
url = 'https://msmarco.z22.web.core.windows.net/msmarcoranking/collection.tar.gz'
drive_folder = 'msmarco_dataset'  # Folder inside Google Drive
filename = 'collection.tar.gz'  # Name of the dataset file

# Execute download and extraction
download_and_extract_to_drive(url, drive_folder, filename)


Mounted at /content/drive
collection.tar.gz already exists in Google Drive.
collection.tar.gz is already extracted in Google Drive.
Dataset is ready at: /content/drive/MyDrive/msmarco_dataset/collection


## Necessary libraries
This cell imports all required Python libraries and modules for the project.
These libraries cover text preprocessing, performance optimization,
file handling, and evaluation metrics for information retrieval.
Key Libraries:

- `nltk` for natural language processing {stemming (reducing words to their root form), stopwords(For removing common words like "the", "is", etc.)}.
- `wordninja` For splitting concatenated words into smaller components
- `re` Regular expressions for pattern matching in text
-`unicodedata` For text normalization (e.g., converting characters to consistent form)
- `pyterrier` For information retrieval tasks (search engines, ranking)
-`ir_measures` For evaluation metrics in information retrieval
- `string` For handling and removing punctuation
- `gc` for garbage collection and memory management.
- `tqdm`  For showing progress bars in loops
- `heapq` for working with priority queues (heaps).
- `tarfile` for extracting tar files.
- `os` For interacting with the operating system (e.g., file paths, environment variables)
- `polars`High-performance data processing library, alternative to pandas
- `matplotlib.pyplot` for plotting and visualizing data.
- `collections` (defaultdict, Counter) for efficient data structures.



In [2]:
#Installing required packages
!pip install wordninja pyterrier ir_measures polars

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 541.6/541.6 kB 7.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 kB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 287.9/287.9 kB 18.5 MB/s eta 0:00:00
  Created wheel for wordninja: filename=wordninja-2.0.0-py3-none-any.whl size=541530 sha256=b9b8c7cffcf314b936fdd6684381e90ea28d4487d88423e3c54c06d8ece7af4c
  Stored in directory: /root/.cache/pip/wheels/e6/66/9c/712044a983337f5d44f90abcd244bd4b8ad28ee64750404b50
Successfully built wordninja


In [3]:
# --- 1. Text Processing and Natural Language Processing (NLP) ---
import re
import string
import nltk
from nltk.stem import SnowballStemmer
from nltk.corpus import stopwords
nltk.download('stopwords', quiet=True)
import unicodedata
import wordninja

# --- 2. Data Structures and Utilities ---
from collections import defaultdict, Counter
import heapq
import math

# --- 3. File Handling and Data Storage ---
import pandas as pd
import polars as pl

# --- 4. Performance and Memory Management ---
import gc
from tqdm import tqdm
from functools import lru_cache

# --- 5. Information Retrieval (IR) ---
import pyterrier as pt
import ir_measures

# --- 6. Visualization and Plotting ---
import matplotlib.pyplot as plt


##Memory Management
###Garbage Collection (gc):

The gc module helps in automatic memory management by clearing unused memory objects. This is important for systems dealing with large datasets like an IR system, where memory usage can grow quickly.



In [4]:
gc.enable()

## Preprocessing

In this cell, we perform several key steps to clean, normalize, and prepare the input text for further analysis in our Information Retrieval (IR) system. The design of these steps was inspired by examining 5 random batches of raw data. Below is a sample of the raw data followed by a detailed explanation of the preprocessing steps implemented in the `preprocess` function.

### Sample Batch of Raw Data

**Batch 1:**

- **PID:** 5216351  
  **Text:** "This difference between the automatic transmission and CVT is the number of gears. The automatic is limited to 4 to 9 gear ratios and there are definite or noticeable changes between the gears and this is felt during driving. The CVT has what is termed an infinite number of ratios ranging from low to high and the transmission moves through the gears steeplessly according to the driving situation."

- **PID:** 2463607  
  **Text:** "WILDLIFE Wildlife at Anedodi is abundant and ever surprising! Being fairly remote and surrounded mainly by large oaks, dogwoods, pine, and other forest tree varieties, the possibilities are endless!"

- **PID:** 2214634  
  **Text:** "A convenient way to access money. Traveller's cheques provide a convenient way to access your money when travelling. By cashing them as you need them, you also minimize the amount of cash you carry around. Denomination Availability. Availability may vary and result in your order not being processed exactly as requested."

- **PID:** 1398422  
  **Text:** "Soldier, Kansas. Soldier is a city in Jackson County, Kansas, United States. As of the 2010 census, the city population was 136."

- **PID:** 3389194  
  **Text:** "Salsa is primarily Cuban son, itself a fusion of Spanish cancion canción and guitar And-Afro Cuban, percussion merged with North American music styles such as Jazz. Salsa also occasionally incorporates elements of Rock & Roll and funk. The term salsa was initially promoted and marketed in New York City during the 1970s. Salsa comprises various musical genres including the Cuban son montuno, guaracha, chachacha, mambo and to a certain extent, bolero."
---

#### 1. **Normalization and Encoding:**
- **Unicode Normalization (NFKD)**: We apply **NFKD (Normalization Form KD)** using `unicodedata.normalize("NFKD", text)`. This decomposes characters into their base form and separates diacritical marks. For example, "é" (a single composed character) is converted into "e" + "´" (base letter + accent). This helps in standardizing text for further processing.
- **Encoding**: The text is then encoded into ASCII using `.encode('ASCII', 'ignore')`, which removes any non-ASCII characters, such as emojis or special symbols, that may cause issues during analysis.
- **Decoding**: After encoding, we decode the text back to UTF-8 using `.decode("utf-8")` to maintain compatibility with further processing steps.
- **Lowercasing**: The text is converted to lowercase using `.lower()` to ensure consistency, as case differences (e.g., "Hello" vs. "hello") should not affect word matching.

#### 2. **Removing Unwanted Patterns (Acronyms, URLs, and Emails):**
- **Acronyms**: The regex pattern `\b(?:[A-Za-z]\.){2,}` matches acronyms (e.g., "U.S.A.") and removes the periods by replacing them with empty strings.
- **URLs**: The patterns `http\S+` and `www\S+` match URLs starting with "http" and "www" respectively. These are replaced with spaces to remove them from the text.
- **Emails**: The pattern `\S+@\S+` matches email addresses and replaces them with spaces.

#### 3. **Removing Punctuation:**
- **Punctuation Removal**: The translation table created by `str.maketrans()` replaces all punctuation characters (from `string.punctuation`) with spaces. This helps eliminate unnecessary symbols that don't carry meaning for many text processing tasks.

#### 4. **Tokenization, Splitting Long Words, and Stopword Removal:**
- **Tokenization**: The text is split into individual words using the `split()` function. This makes it easier to work with each word individually.
- **Avoiding Consecutive Duplicates**: We use the `current` variable to avoid adding consecutive duplicates. This ensures that the same word doesn't appear twice in a row due to tokenization or splitting.

 ***Example:***  
  For the text:  
  `"hello hello world world hello"`  
  The process will yield:  
  `["hello", "world", "hello"]`  
  (Note that only consecutive duplicates are removed.)

#### 5. **Splitting Long Words (If Necessary):**
- **Splitting Long Words**: If a word is longer than 10 characters, we use the `split_word()` function (from the `wordninja` library) to split it into smaller, more manageable sub-words. This is especially useful for handling compound words that are concatenated without spaces.

 ***Example:***  
  Consider the long concatenated word:  
  `"thisisaverylongcompoundword"`  
  Using `wordninja.split(word)`, it might be split into:  
  `["this", "is", "a", "very", "long", "compound", "word"]`  
  This makes the text easier to process and match against the lexicon.


#### 6. **Stopword Removal and Stemming:**
- **Stopwords Removal**: We check each word (or sub-word, if it was split) against a list of common "stopwords" (such as "the", "and", "is"). These are typically removed because they do not contribute much meaningful information in text analysis.
- **Stemming**: After removing stopwords, we apply stemming using the `SnowballStemmer` from NLTK. Stemming reduces words to their root form (e.g., "running" becomes "run"). This allows us to treat different forms of the same word as equivalent.

#### **Output:**
The final output of the `preprocess` function is a list of **preprocessed words**. This list consists of tokens that have been cleaned, normalized, tokenized, split, and stemmed, with stopwords removed.


In [5]:
# ----------------------------------------------------
# 1. Precompile regex patterns and initialize globals
# ----------------------------------------------------
acronym_regex = re.compile(r"\b(?:[A-Za-z]\.){2,}")
url_http_regex = re.compile(r"\s?http\S+")
url_www_regex = re.compile(r"\s?www\S+")
email_regex = re.compile(r"\S+@\S+")
punct_trans_table = str.maketrans(string.punctuation, " " * len(string.punctuation))

# Cache the stopwords and stemmer as globals so they are not re-created on every call.
stopwords_set = set(stopwords.words("english"))
stemmer = SnowballStemmer("english")

# ----------------------------------------------------
# 2. Define the helper functions using the precompiled objects
# ----------------------------------------------------
def split_word(word):
    """Splits a long word into smaller components using wordninja."""
    return wordninja.split(word)

def preprocess(text):
    """
    Cleans, normalizes, tokenizes, and stems the input text.
    Returns:
        list[str]: Preprocessed and stemmed tokens.
    """
    # Normalize and encode text
    text = unicodedata.normalize("NFKD", text).encode('ASCII', 'ignore').decode("utf-8").lower()

    # Remove acronyms (e.g., "U.S.A.")
    text = acronym_regex.sub(lambda x: x.group().replace('.', ''), text)

    # Remove URLs and emails
    text = url_http_regex.sub(' ', text)
    text = url_www_regex.sub(' ', text)
    text = email_regex.sub(' ', text)

    # Remove punctuation (by replacing punctuation with spaces)
    text = text.translate(punct_trans_table)

    words = text.strip().split()
    words_with_no_close_duplicate = []
    current = ""

    for word in words:
        if word != current:  # Skip consecutive duplicates
            # Split long words into subwords if needed
            splitted_words = split_word(word) if len(word) > 10 else [word]
            for splitted_word in splitted_words:
                if splitted_word not in stopwords_set:
                    words_with_no_close_duplicate.append(stemmer.stem(splitted_word))
            current = word

    return words_with_no_close_duplicate

#**IR Data Structures**

## **Index Construction**
In this section, the `build_index` function constructs an **inverted index** and other supporting data structures for an **Information Retrieval (IR) system**. The implementation is **modular**, with clearly defined responsibilities across multiple functions to improve **readability, maintainability, and performance**.

## **Overview of the Indexing Process**  
The indexing process involves:  

- **Lexicon Construction**: Assigning unique term IDs to words and tracking their frequency.  
- **Direct Indexing**: Mapping documents to the terms they contain.  
- **Inverted Indexing**: Mapping terms to the documents they appear in.  
- **Document Statistics**: Storing document lengths and token counts.  

This implementation efficiently processes a dataset containing **documents** (`docno`) and their **tokens** (preprocessed words).  



### **Initializing Index Data Structures**  
### **Purpose:**  
This function **creates and initializes** all required data structures:  

- **`lexicon`**: Maps tokens (terms) to `[termid, document frequency, term frequency]`.  
  - `termid`: A unique identifier for each term.  
  - `document frequency`: Number of documents containing the term.  
  - `term frequency`: Total occurrences of the term across all documents.  
- **`doc_index`**: Stores document lengths as tuples `(docid, doclen)`, where `doclen` is the number of tokens in the document.  
- **`direct_index`**: Maps document IDs to term IDs and their term frequencies.  
- **`inv_d` & `inv_f`**: Store **inverted index** data:  
  - **`inv_d`**: Maps term IDs to document IDs.  
  - **`inv_f`**: Maps term IDs to their frequency in each document.

In [6]:
def setup_index_structures():
    """
    Initializes the required data structures for indexing.

    Returns:
        tuple: Empty structures for lexicon, document index, direct index, and inverted index.
    """
    lexicon = defaultdict(lambda: [0, 0, 0])  # [termid, document frequency, term frequency]
    doc_index = []
    direct_index = defaultdict(dict)
    inv_d, inv_f = defaultdict(list), defaultdict(list)  # Inverted index storage
    return lexicon, doc_index, direct_index, inv_d, inv_f


### **Lexicon Management**  
### **Purpose:**  
- Assigns a **unique ID** (`termid`) to each token.  
- Uses `termid_counter` (a **mutable list**) to maintain a **global term ID counter**.  

### **Why use `termid_counter` as a list?**  
- Python **integers are immutable**, so they **cannot** be modified inside a function.  
- Using a **list** (`termid_counter[0]`) allows modifying it **in-place**.

In [7]:
def update_lexicon(token, lexicon, termid_counter):
    """
    Updates the lexicon with a new token if it does not exist.

    Parameters:
        token (str): The token to be added.
        lexicon (dict): The lexicon mapping tokens to metadata.
        termid_counter (list[int]): Mutable counter for term IDs.

    Returns:
        int: The term ID of the token.
    """
    if token not in lexicon:
        lexicon[token][0] = termid_counter[0]
        termid_counter[0] += 1
    return lexicon[token][0]


### **Processing Individual Documents**
### **Purpose:**  
- **Counts token frequencies** in the document using `Counter`.  
- Calls `update_lexicon()` to assign a term ID.  
- Updates the followings:  
  - **Inverted Index** (`inv_d`, `inv_f`).  
  - **Direct Index** (`direct_index`).  
  - **Lexicon Statistics** (`lexicon`).  
- Returns **document length** (`doclen`).  

### **Data Structures Updated:**  
- **`lexicon`**: Maps tokens (terms) to `[termid, document frequency, term frequency]`.  
- **`direct_index`**: Stores term-document mappings.  
- **`inv_d`**: Stores document occurrences per term.  
- **`inv_f`**: Stores term frequency per document.  

In [8]:
def process_document(docid, tokens, lexicon, direct_index, inv_d, inv_f, termid_counter):
    """
    Processes a single document, updating all index structures.

    Parameters:
        docid (int): The document identifier.
        tokens (list[str]): The list of tokens in the document.
        lexicon (dict): Lexicon mapping tokens to term IDs and statistics.
        direct_index (dict): Direct index mapping doc IDs to term frequencies.
        inv_d (dict): Inverted index storing document occurrences per term.
        inv_f (dict): Inverted index storing term frequencies per document.
        termid_counter (list[int]): Mutable counter for assigning term IDs.

    Returns:
        int: The length of the document (number of tokens).
    """
    token_tf = Counter(tokens)  # Count term frequencies within the document

    for token, tf in token_tf.items():
        token_id = update_lexicon(token, lexicon, termid_counter)

        # Update inverted and direct index
        inv_d[token_id].append(docid)
        inv_f[token_id].append(tf)
        direct_index[docid][token_id] = tf

        # Update lexicon statistics
        lexicon[token][1] += 1  # Document frequency
        lexicon[token][2] += tf  # Term frequency

    return len(tokens)


### **Main Index Construction**
### **Purpose:**  
- **Initializes** all indexing structures.  
- Iterates over each document using `tqdm` for **progress tracking**.  
- Calls `process_document()` for each document's **tokens**.  
- Updates **document statistics** (`num_docs`, `num_terms`, `num_tokens`).  

### **Returns:**  
- **`lexicon`**: Term metadata.  
- **`inverted_index`**: `{docids, freqs}`.  
- **`doc_index`**: Document lengths.  
- **`direct_index`**: Term-document mappings.  
- **`stats`**: Summary statistics.  

In [9]:
def build_index(dataset):
    """
    Constructs an inverted index and other supporting structures from the input dataset.

    Parameters:
        dataset (DataFrame): A pandas DataFrame with the following columns:
            - 'docno': Unique identifier for each document.
            - 'tokens': List of tokens for the document.

    Returns:
        tuple: A collection of indices and statistics:
            - lexicon (dict): Maps tokens to their metadata [termid, document frequency, term frequency].
            - inverted_index (dict): Contains document IDs ('docids') and term frequencies ('freqs').
            - doc_index (list): List of tuples (docid, doclen), where doclen is the number of tokens in each document.
            - direct_indexfreqs (dict): Maps document IDs to term IDs and their term frequencies.
            - stats (dict): Summary statistics.
    """
    lexicon, doc_index, direct_index, inv_d, inv_f = setup_index_structures()
    termid_counter = [0]  # Using a mutable list for consistent term ID tracking
    total_tokens = 0

    for row in tqdm(dataset.itertuples(index=False), desc="Indexing", total=len(dataset)):
        docid = int(row.docno)
        tokens = row.tokens
        doclen = process_document(docid, tokens, lexicon, direct_index, inv_d, inv_f, termid_counter)
        doc_index.append((docid, doclen))
        total_tokens += doclen

    stats = {
        "num_docs": len(dataset),
        "num_terms": len(lexicon),
        "num_tokens": total_tokens,
    }

    return dict(lexicon), {"docids": dict(inv_d), "freqs": dict(inv_f)}, doc_index, dict(direct_index), stats


## **InvertedIndex**

The `InvertedIndex` class represents an **inverted index** used in information retrieval systems.  
It allows **efficient document retrieval** based on term queries and supports the **BM25 ranking function**.

### Key Responsibilities
- Stores the mapping of terms to documents (**Lexicon** & **Inverted Index**).
- Manages document metadata (**Document Index**).
- Implements **BM25 ranking** to score documents based on query terms.
- Provides efficient **posting list iteration** for search operations.

### Main Components
- **Lexicon (`lexicon`)**: Maps terms to metadata like term IDs and frequencies.
- **Inverted Index (`inv`)**: Maps term IDs to **posting lists** of documents.
- **Document Index (`doc`)**: Stores document IDs and their lengths.
- **Direct Index (`direct`)**: Maps document IDs to term frequencies.
- **Statistics (`stats`)**: Stores global collection stats (e.g., total number of documents).
- **BM25 Parameters (`K`, `B`)**: Used for document ranking.





### **Nested Class: `PostingListIterator`**
The `PostingListIterator` class is **a helper class inside `InvertedIndex`** that iterates over a term's posting list.

### Why is this useful?
- Allows **efficient traversal** of documents containing a specific term.
- Computes **BM25 scores** for ranking documents.
- Provides functions to **jump** to a target document efficiently.

### Attributes
| Attribute | Description |
|-----------|------------|
| `docids` | List of document IDs where the term appears. |
| `freqs` | Term frequencies in each document. |
| `doclens` | A dictionary mapping document IDs to document lengths. |
| `N` | Total number of documents in the collection. |
| `avgdl` | Average document length. |
| `K, B` | BM25 parameters for controlling term saturation and length normalization. |
| `pos` | Current position in the posting list. |


####Methods

## `docid()`: Get Current Document ID
**Purpose**:  
Returns the document ID at the current iterator position.

**Returns**:  
- `int`: Document ID.
- `math.inf`: If the iterator reaches the end.

---

###**`score()`: Compute BM25 Score**
**Purpose**:  
Computes the **BM25 relevance score** of the current document.

**Formula**:  

$$
BM25(i) = \log \frac{N}{df_i} \times \frac{tf_i}{K \cdot ((1 - B) + B \cdot dl_i / avgdl) + tf_i}$$



Where:  
- \( tf_i \) = Term frequency in the document.  
- \( df_i \) = Document frequency (number of documents containing the term).  
- \( dl_i \) = Length of the current document.  
- \( avgdl \) = Average document length in the collection.  
- \( N \) = Total number of documents.  
- \( K, B \) = BM25 parameters.  

**Returns**:  
- `float`: BM25 score.  
- `math.inf`: If iterator reaches the end.


---

### **`next(target)`: Move Iterator**
**Purpose**:  
Moves the iterator:
- To the **next document** if no target is given.
- To a **specific document ID** if `target` is provided.

**Parameters**:  
- `target` *(int, optional)*: The document ID to jump to.

**Efficiency**:  
If `target` is given, it **skips over documents**, making retrieval faster.

---


###**`is_end_list()`**:

**Purpose**:

  - Checks if the iterator has reached the end of the posting list.

---

###**`len()`**:
**Purpose**:
   - Returns the total number of documents in the posting list.



###**`InvertedIndex` Class Constructor**

**Purpose:**
Initializes an instance of the InvertedIndex class with the provided lexicon, inverted index, document index, direct index, statistics, and BM25 parameters (K, B).

## Attributes
| Attribute | Description |
|-----------|------------|
| `lexicon` | A dictionary mapping terms to metadata like term IDs and frequencies. |
| `inv` |  An inverted index containing term IDs and posting lists (document IDs and term frequencies). |
| `doc` |  A list of documents, each containing a document ID and length. `(docid, doclen)`. |
| `direct` | Direct index: Maps document IDs to term frequencies. |
| `stats` | A dictionary containing collection statistics like the total number of documents. |
| `N` | Total number of documents. |
| `doclens` | A dictionary mapping document IDs to document lengths. |
| `avgdl` | Average document length in the collection. |
| `K, B` | BM25 scoring parameters. |



####**Methods**
###**`num_docs()`**
**Purpose**:  
Returns the **total number of documents** in the index.

**Why is this important?**  
- Used in **BM25 scoring** to compute the **inverse document frequency (IDF)**.
- Helps in **normalizing** scores based on collection size.

**Returns**:  
- `int`: Total number of documents.

###**`get_posting(termids)`**
**Purpose**:  
Fetches the **posting list** of a given term.

**How does it work?**  
- Looks up the **term ID** in `lexicon`.
- Retrieves document IDs & term frequencies from `inv` index.
- Returns a `PostingListIterator` for **efficient iteration**.

**Returns**:  
- `PostingListIterator`: Iterator for the posting list `list[PostingListIterator]`.
- `[]`: If the term does not exist.

### **`get_termids(tokens)`**
**Purpose**:  
Retrieves the term IDs for a list of tokens (terms).

**How does it work?**  
- For each token in the tokens list (of the query), it checks whether the token exists in the lexicon.
-If found, it retrieves the term ID associated with the token.
-Returns a list of term IDs for the tokens.

**Returns**:  
- `list[int]`: A list of term IDs corresponding to the tokens in the input list.
- `[]`: If a token is not found in the lexicon.

### **`get_postings(termids)`**:
   - Retrieves a list of posting lists for the provided term IDs.
   - Returns a list of `PostingListIterator` objects, one for each term ID.


In [10]:
class InvertedIndex:
    """
    Represents an inverted index for document retrieval.

    This class provides methods to manage the lexicon, inverted index, document index, and retrieval statistics.
    It also supports BM25 scoring for ranking documents based on term relevance.

    Attributes:
        lexicon (dict): Mapping of terms to metadata (e.g., term ID, frequencies).
        inv (dict): Inverted index containing 'docids' and 'freqs'.
        doc (list): Document index containing document IDs and their lengths.
        direct (dict): Direct index mapping document IDs to term IDs and their frequencies.
        stats (dict): Index statistics, including number of documents.
        N (int): Total number of documents in the collection.
        avgdl (float): Average document length.
        K (float): BM25 parameter for term frequency scaling.
        B (float): BM25 parameter for document length normalization.
    """

    class PostingListIterator:
        """
        Iterates over a posting list for a given term, supporting BM25 scoring.

        Attributes:
            docids (list): List of document IDs containing the term.
            freqs (list): Term frequencies in each document.
            doclens (dict): Mapping of document IDs to their lengths.
            N (int): Total number of documents.
            avgdl (float): Average document length.
            K (float): BM25 parameter for term frequency scaling.
            B (float): BM25 parameter for document length normalization.
        """
        def __init__(self, docids, freqs, doclens, N, avgdl, k, b):
            self.docids = docids
            self.freqs = freqs
            self.doclens = doclens
            self.N = N
            self.avgdl = avgdl
            self.K = k
            self.B = b
            self.pos = 0  # Current position in the posting list


        def docid(self):
            """
            Returns the document ID at the current iterator position.

            Returns:
                int: The current document ID, or math.inf if the list has ended.
            """
            if self.is_end_list():
                return math.inf
            return self.docids[self.pos]

        def score(self):
            """
            Computes the BM25 score for the current document.

            Returns:
                float: The BM25 score for the document, or math.inf if the list has ended.
            """
            if self.is_end_list():
                return math.inf

            tf_i = self.freqs[self.pos]
            dl_i = self.doclens.get(self.docid())
            df_i = self.len()

            return 0 if df_i == 0 else (tf_i / (self.K * ((1 - self.B) + self.B * dl_i / self.avgdl) + tf_i)) * math.log(self.N / df_i)

        def next(self, target=None):
            """
            Advances the iterator to the next position or to a target document ID.

            Parameters:
                target (int, optional): The target document ID to jump to. Defaults to None.
            """
            if not target:
                if not self.is_end_list():
                    self.pos += 1
            else:
                if target > self.docid():
                    try:
                        self.pos = self.docids.index(target, self.pos)
                    except ValueError:
                        self.pos = len(self.docids)

        def is_end_list(self):
            """
            Checks if the iterator has reached the end of the posting list.

            Returns:
                bool: True if the list has ended, False otherwise.
            """
            return self.pos == len(self.docids)

        def len(self):
            """
            Returns the length of the posting list.

            Returns:
                int: The number of documents in the posting list.
            """
            return len(self.docids)

    def __init__(self, lex, inv, doc, direct, stats, k, b):
        self.lexicon = lex
        self.inv = inv
        self.doc = doc
        self.direct = direct
        self.stats = stats
        self.N = self.num_docs()
        self.doclens = {int(doc_entry[0]): doc_entry[1] for doc_entry in self.doc}
        total_dl = sum(doc_entry[1] for doc_entry in self.doc)
        self.avgdl = total_dl / self.N
        self.K = k
        self.B = b

    def num_docs(self):
        """
        Returns the total number of documents.

        Returns:
            int: Total document count.
        """
        return self.stats['num_docs']

    def get_posting(self, termid):
        """
        Retrieves the posting list iterator for a given term ID.

        Parameters:
            termid (int): The ID of the term.

        Returns:
            PostingListIterator: An iterator for the term's posting list.
        """
        docids = self.inv['docids'].get(termid)
        freqs = self.inv['freqs'].get(termid)

        return InvertedIndex.PostingListIterator(
            docids=docids,
            freqs=freqs,
            doclens=self.doclens,
            N=self.N,
            avgdl=self.avgdl,
            k=self.K,
            b=self.B
        )

    def get_termids(self, tokens):
        """
        Retrieves the term IDs for a list of tokens.

        Parameters:
            tokens (list[str]): List of tokens.

        Returns:
            list[int]: List of term IDs.
        """
        return [self.lexicon[token][0] for token in tokens if token in self.lexicon]

    def get_postings(self, termids):
        """
        Retrieves the posting list iterators for a list of term IDs.

        Parameters:
            termids (list[int]): List of term IDs.

        Returns:
            list[PostingListIterator]: List of posting list iterators for the terms.
        """
        return [self.get_posting(termid) for termid in termids]


# DAAT Algorithm

The Document-At-A-Time (DAAT) algorithm processes multiple posting lists in parallel, advancing through each list simultaneously to identify candidate documents for multi-term queries. This approach allows for early elimination of non-relevant documents and minimizes unnecessary comparisons.

- Support for Boolean and Phrasal Queries
- Efficiency for skipping irrilevance Documents

## Components of DAAT

### **Function: `get_min_docid`**
In an inverted index, each posting list is constructed by scanning through the documents, typically in the order of document IDs. As a result, the document IDs in each posting list are stored in sorted order. This ordering is essential because it allows for efficient merging and retrieval of documents when processing queries.

The `get_min_docid` function leverages the sorted order by iterating over each posting list iterator and retrieving the current document ID (using the `docid()` method) from those lists that still have pending documents. Because the posting lists are sorted, the smallest pending document ID among all iterators is chosen as the next candidate for processing.This mechanism enables a **single-pass traversal** of the posting lists, making the retrieval process efficient and ensuring each document is processed only once.

### **Parameters**
| Parameter  | Type                              | Description |
|------------|-----------------------------------|-------------|
| `postings` | `list[PostingListIterator]`      | A list of posting list iterators, each providing access to a document list associated with a term. |

### **Returns**
| Return Type | Description |
|-------------|-------------|
| `int`       | The smallest pending document ID (`docid`) among the posting list iterators. Returns `math.inf` if all posting lists are finished. |

### **Functionality**
1. **Initialize `min_docid` to `math.inf`**: This ensures that any valid document ID encountered will be smaller than the initial value.
2. **Iterate through each posting list iterator (`p`)**:
   - If the iterator is **not finished** (`p.is_end_list()` is `False`):
     - Retrieve the current `docid` using `p.docid()`.
     - Update `min_docid` with the smaller value between `p.docid()` and `min_docid`.
3. **Return `min_docid`**:
   - If at least one posting list contains pending documents, the function returns the smallest `docid`.
   - If all posting lists are finished, it returns `math.inf`, signaling that no documents remain.

### **Edge Cases Considered**
- If all posting lists are **empty or fully traversed**, the function returns `math.inf`.  
- Handles scenarios where some posting lists are **still pending while others are finished**.  
- Efficiently finds the **smallest pending document ID** in **O(n) time complexity**, where `n` is the number of posting list iterators.  
- This function is a key component in **merging multiple posting lists efficiently** for query processing in an **inverted index-based search engine**.  


In [11]:
def get_min_docid(postings):
    """
    Finds the smallest pending document ID (docid) among the provided postings.

    Parameters:
        postings (list[PostingListIterator]): List of posting list iterators.

    Returns:
        int: The smallest pending document ID, or math.inf if all postings are finished.
    """
    min_docid = math.inf
    for p in postings:
        if not p.is_end_list():  # Check if the posting list has not been fully traversed
            min_docid = min(p.docid(), min_docid)
    return min_docid

### **TopKPriorityQueue Class**

The `TopKPriorityQueue` class implements a priority queue designed to maintain the top-k elements based on their scores. This data structure is essential for efficiently managing and updating a limited set of the highest-scoring items, commonly used in information retrieval algorithms like Document-At-A-Time (DAAT).

## **Attributes**
The following table describes the attributes of the `TopKPriorityQueue` class:

| Attribute  | Type  | Description  |
|------------|-------|--------------|
| `queue`    | `list` | A min-heap that stores tuples of `(score, docid)`. The smallest score in the heap is at the root. |
| `k`        | `int` | The maximum capacity of the queue, determining how many top items can be stored. |
| `threshold` | `float` | The minimum score required for an element to be considered for insertion into the queue. |

---

## **Methods**

### **1. `__init__(self, k=10, threshold=0.0)`**
**Purpose:**  
Initializes the `TopKPriorityQueue` instance with a given maximum capacity (`k`) and an optional threshold.

**Parameters:**
- `k` (`int`): Maximum size of the queue. Defaults to 10.
- `threshold` (`float`): Initial threshold value. Defaults to 0.0.

---

### **2. `size(self)`**
**Purpose:**  
Returns the current number of elements in the queue.

**Returns:**
- `int`: The number of items currently in the queue.

---

### **3. `is_eligible(self, score)`**
**Purpose:**  
Determines whether a given score exceeds the current threshold and could be added to the queue.

**Parameters:**
- `score` (`float`): The score to evaluate against the threshold.

**Returns:**
- `bool`: `True` if the score exceeds the threshold, otherwise `False`.

---

### **4. `reset(self, new_threshold=None)`**
**Purpose:**  
Clears the queue and optionally updates the threshold.

**Parameters:**
- `new_threshold` (`float`, optional): If provided, updates the queue’s threshold to this new value.

**Behavior:**
- Resets the queue to an empty list.
- Updates the `threshold` attribute if a new value is provided.

---

### **5. `__repr__(self)`**
**Purpose:**  
Provides a string representation of the `TopKPriorityQueue` instance for debugging and logging purposes.

**Returns:**
- `str`: A formatted string describing the queue's current state.

**Example Output:**
<6 items, th=4.0 [(4.0, 1), (5.0, 2), (9.0, 3)]>


---

### **6. `add(self, docid, score)`**
**Purpose:**  
Attempts to insert a `(docid, score)` pair into the queue. The insertion only occurs if the score exceeds the threshold.

**Parameters:**
- `docid` (`int`): The unique identifier for the document.
- `score` (`float`): The score associated with the document.

**Returns:**
- `bool`: `True` if the item was successfully inserted, `False` otherwise.

**Behavior:**
1. If `score` is greater than `threshold`:
   - If the queue has reached its maximum capacity (`k`), replace the root element (smallest score) using `heapq.heapreplace` if the new score is higher.
   - If the queue is not full, add the new `(score, docid)` pair using `heapq.heappush`.
2. Updates the `threshold`:
   - If the queue is full, the threshold is set to the smallest score in the queue (`queue[0][0]`).

---

## **Heap Behavior**
The `queue` is maintained as a **min-heap**, ensuring that:
- The smallest score is always at the root.
- Insertions and deletions operate efficiently, maintaining the top-k elements.


In [12]:
class TopKPriorityQueue:
    """
    A priority queue to manage the top-k elements with a dynamic threshold.

    Attributes:
        queue (list): The internal priority queue storing (score, docid) tuples.
        k (int): Maximum capacity of the queue.
        threshold (float): Current threshold to determine if a new element enters the queue.
    """

    def __init__(self, k=10, threshold=0.0):
        """
        Initializes the queue with a maximum capacity and an initial threshold.

        Parameters:
            k (int): Maximum size of the queue. Defaults to 10.
            threshold (float): Initial threshold value. Defaults to 0.0.
        """
        self.queue = []
        self.k = k
        self.threshold = threshold

    def get_size(self):
        """
        Returns the number of elements in the queue.

        Returns:
            int: The current size of the queue.
        """
        return len(self.queue)

    def is_eligible(self, score):
        """
        Checks if a score exceeds the current threshold.

        Parameters:
            score (float): The score to evaluate.

        Returns:
            bool: True if the score exceeds the threshold, False otherwise.
        """
        return score > self.threshold

    def reset(self, new_threshold=None):
        """
        Clears the queue and optionally updates the threshold.

        Parameters:
            new_threshold (float, optional): The new threshold value. Defaults to None.
        """
        self.queue = []
        if new_threshold is not None:
            self.threshold = new_threshold

    def __repr__(self):
        """
        Returns a string representation of the queue.

        Returns:
            str: A string describing the queue's state.
        """
        return f'<{self.get_size()} items, th={self.threshold} {self.queue}>'

    def add(self, docid, score):
        """
        Inserts an element into the queue if the score exceeds the threshold.

        Parameters:
            docid (int): Document ID associated with the score.
            score (float): The score to insert.

        Returns:
            bool: True if the element was inserted, False otherwise.
        """
        if score > self.threshold:
            if self.get_size() >= self.k:
                heapq.heapreplace(self.queue, (score, docid))
            else:
                heapq.heappush(self.queue, (score, docid))
            if self.get_size() >= self.k:
                self.threshold = max(self.threshold, self.queue[0][0])
            return True
        return False


##  Algorithm Implementation

### Purpose:
The `daat` function implements the **Document-At-A-Time (DAAT)** retrieval algorithm. Its goal is to combine scores from multiple posting lists—each representing documents containing a specific query term—into a single ranked list of documents. DAAT is used because it allows for efficient, simultaneous processing of multiple posting lists, making it well-suited for handling large-scale document collections. By iterating over the documents across posting lists in parallel, DAAT minimizes redundant work and quickly identifies the most relevant documents. This approach is particularly advantageous when working with ranking models like BM25, as it aggregates term scores on a per-document basis, ensuring that the top-k scoring documents are efficiently retrieved.

### Parameters:
- **postings** (list of `PostingListIterator`):  
  A list of posting list iterators, where each iterator corresponds to a term’s posting list.  
  Each `PostingListIterator` provides:
  - `docid()`: Returns the current document ID in the posting list.
  - `score()`: Computes the BM25 score (or similar scoring function) for the current document.
  - `next()`: Advances the iterator to the next document.
  - `is_end_list()`: Checks if the iterator has reached the end of the posting list.

- **k** (int, optional):  
  The maximum number of top results to return.  
  Default value is `10`. This determines the capacity of the top-k result set maintained during execution.

### Internal Attributes / Variables:
- **top**:  
  An instance of `TopKPriorityQueue` (a priority queue) initialized with capacity `k`. It is used to store the top-k document IDs along with their aggregated scores.

- **current_docid**:  
  The smallest document ID among all the posting lists at the current iteration, obtained via a helper function (e.g., `min_docid(postings)`). This serves as the current candidate document for scoring.

- **score**:  
  A cumulative score computed for the current document by summing individual scores from all posting lists where the document appears.

- **next_docid**:  
  A variable to track the next smallest document ID across all posting lists. It is updated during each iteration to determine the next document to process.

### Returns:
- **List of tuples (score, docid)**:  
  The function returns a list where each tuple contains:
  - **score**: The aggregated BM25 score (or equivalent metric) for a document.
  - **docid**: The unique identifier of the document.
  
  The list is **sorted in descending order** by score, ensuring that the most relevant documents (with the highest scores) appear first.

### Detailed Workflow:
1. **Initialization**:  
   - A priority queue `top` is created to hold the top-k results.
   - The initial `current_docid` is determined by selecting the smallest document ID across all posting lists using `min_docid(postings)`.

2. **Iterative Processing**:  
   - For the `current_docid`, the function iterates over all posting lists:
     - If a posting list’s current document ID equals `current_docid`, its score is added to the cumulative `score` and its iterator advances.
     - Simultaneously, the smallest document ID from the remaining posting lists (i.e., `next_docid`) is determined.
   
3. **Result Insertion**:  
   - The document and its cumulative score `(current_docid, score)` are inserted into the `top` priority queue.

4. **Termination and Output**:  
   - The loop continues until all posting lists are finished (i.e., `current_docid` becomes `math.inf`).
   - Finally, the function returns the contents of the `top` queue, sorted in descending order by score.
---

**Usage Example:**  
After processing, the function might return a list like:  
```python
[(23.5, 42), (21.0, 17), (19.8, 3), ...]
```
indicating that document 42 has a score of 23.5, document 17 a score of 21.0, etc.

In [13]:
def daat(postings, k=10):
    """
    Implements the Document-At-A-Time (DAAT) algorithm for retrieving relevant documents.

    Parameters:
        postings (list[PostingListIterator]): List of posting list iterators, each representing a term's posting list.
        k (int): Maximum number of results to return. Defaults to 10.

    Returns:
        list[tuple]: A list of tuples (score, docid), sorted in descending order of score.
    """
    top = TopKPriorityQueue(k)  # Priority queue to store the top-k results
    current_docid = get_min_docid(postings)  # Initialize with the smallest docid across all postings

    while current_docid != math.inf:
        score = 0
        next_docid = math.inf

        # Compute the score for the current document and find the next docid
        for posting in postings:
            if posting.docid() == current_docid:
                score += posting.score()
                posting.next()  # Move to the next position in the posting list
            if not posting.is_end_list():
                next_docid = min(posting.docid(), next_docid)

        # Insert the current document and its score into the top-k queue
        top.add(current_docid, score)

        # Update to the next document ID to process
        current_docid = next_docid

    # Return the top-k results sorted by score in descending order
    return sorted(top.queue, reverse=True)

# Getting Output

## Retrieving Results for query
## Function: retrieve_query_results

### Purpose:
The `retrieve_query_results` function is designed to process a text query and retrieve the most relevant documents from a collection using an inverted index. It combines **query preprocessing**, **mapping of query tokens to term IDs**, **retrieval of posting lists**, and finally, ranking of documents via the **Document-At-A-Time (DAAT)** algorithm. This function essentially serves as the entry point for handling user queries in an Information Retrieval (IR) system.

### Parameters:
- **query** (str):  
  The input text query provided by the user.  
  *Example:* `"What is the capital of France?"`

- **index** (`InvertedIndex`):  
  An instance of the `InvertedIndex` class, which contains the lexicon, posting lists, document index, and other retrieval statistics necessary for scoring documents.  
  *Key Attributes from the index include:*  
  - `lexicon`: A mapping of tokens to their metadata (e.g., term IDs, frequencies).  
  - `inv`: The inverted index storing document IDs and term frequencies.  
  - `doc`: The document index with document lengths.  
  - `stats`: Collection-wide statistics such as the total number of documents.

- **heap_size** (int, optional):  
  The maximum number of top results to return.  
  Defaults to `10`. This parameter sets the capacity for the result set (the top-k documents).

### Internal Workflow and Attributes:
1. **Query Preprocessing**:  
   - The function first applies the `preprocess` function to the input query.  
   - **Output:** A set of cleaned and stemmed tokens (`qtokens`) is generated from the query.  
   -**Purpose:** This normalization ensures that the query is compatible with the indexed tokens and reduces noise.

2. **Mapping Tokens to Term IDs**:  
   - The cleaned query tokens are then converted into their corresponding term IDs using `index.get_termids(qtokens)`.  
   - **Output:** A list of term IDs (`qtermids`) representing the query terms.  
   -**Purpose:** This mapping allows the system to locate the relevant posting lists in the inverted index.

3. **Retrieving Posting Lists**:  
   - The function retrieves posting list iterators for each term ID via `index.get_postings(qtermids)`.  
   - **Output:** A list of `PostingListIterator` objects (`postings`), each representing the documents and term frequencies for a specific query term.  
   - **Purpose:** These posting lists form the basis for scoring documents.

4. **Document Ranking via DAAT**:  
   - Finally, the function calls the `daat` function, passing the retrieved posting lists and the specified `heap_size`.  
   - **Output:** The `daat` function returns a list of tuples `(score, docid)` sorted in descending order by score, representing the top-ranked documents.  
   -**Purpose:** The DAAT algorithm aggregates scores across posting lists and efficiently identifies the most relevant documents for the query.

### Returns:
- **list[tuple]**:  
  A list of tuples where each tuple consists of:
  - **score** (float): The aggregated relevance score (e.g., BM25 score) for the document.
  - **docid** (int): The unique identifier of the document.

  The returned list is sorted in **descending order** by score, ensuring that the most relevant documents appear first.

---

### Summary:
The `retrieve_query_results` function integrates query preprocessing, term mapping, posting list retrieval, and document ranking into a single pipeline. It leverages the capabilities of the `InvertedIndex` class and the DAAT algorithm to efficiently handle user queries and return the top-k relevant documents.


In [14]:
def retrieve_query_results(query, index, heap_size=10):
    """
    Processes a text query and retrieves the most relevant results using an inverted index.

    Parameters:
        query (str): The input text query.
        index (InvertedIndex): The inverted index used to retrieve postings and scores.
        heap_size (int): Maximum number of results to return. Defaults to 10.

    Returns:
        list[tuple]: A list of tuples (score, docid), sorted in descending order of score.
    """
    # Preprocess the query to obtain cleaned and stemmed tokens
    qtokens = set(preprocess(query))

    # Map query tokens to their term IDs
    qtermids = index.get_termids(qtokens)

    # Retrieve postings for the term IDs
    postings = index.get_postings(qtermids)

    # Execute the DAAT algorithm to get the top results
    return daat(postings, heap_size)

## Generating TREC for IR
## Function: generate_trec_run_file  

### Purpose:
The `generate_trec_run_file` function generates a TREC-formatted run file from a set of query. It retrieves the top-k relevant documents for each query using an inverted index (if provided) and then formats the results according to the TREC run file specification. This file can be used for evaluation in standard IR benchmarks.

### Parameters:
- **queries_df** (DataFrame):  
  A pandas DataFrame containing the query. It must include at least the following columns:
  - `query`: The text of the query.
  - `qid`: The unique query identifier.
  
- **inv_index** (InvertedIndex or None):  
  An instance of the `InvertedIndex` class used for retrieving document scores based on the query.  
  - If provided, the function uses `retrieve_query_results` to generate the top-k results.
  - If `None`, it assumes that the `queries_df` DataFrame already contains precomputed results (with columns like `docno`, `rank`, and `score`).

- **file_name** (str):  
  The name (and optionally the path) of the output file where the TREC run file data will be saved.

### Behavior and Internal Workflow:
1. **Set Retrieval Size**:  
   - A constant `k` is set to `10`, meaning the function will retrieve up to 10 results per query.

2. **Result Generation (if inv_index is provided)**:  
   - For each query in the `queries_df` DataFrame:
     - The query text is processed by `retrieve_query_results` to obtain a list of results, where each result is a tuple `(score, docid)`.
     - The function pairs each result with the query ID and formats it according to the TREC run file format:  
       `<query_id> Q0 <docno> <rank> <score> run.txt`
     - All formatted strings are collected into a list called `results`.

3. **Result Generation (if inv_index is None)**:  
   - If no inverted index is provided, the function assumes that the `queries_df` DataFrame already contains the result details.
   - For each row, it directly uses the provided `qid`, `docno`, `rank` (starts from 1 then decrease incrementally), and `score` values to format the result string.

4. **File Output**:  
   - The list of formatted result strings is written to the specified file (`file_name`) using standard file I/O operations.

### Returns:
- **None**:  
  The function writes the formatted run results directly to the specified file. It does not return any value, but its output is the generated run file in TREC format.

---



In [15]:
def generate_trec_run_file(queries_df, inv_index, file_name):
    """
    Generates a TREC run file from queries_df and an inverted index.

    Parameters:
        queries_df (DataFrame): DataFrame containing queries with columns 'query' and 'qid'.
        inv_index (InvertedIndex or None): Inverted index used to retrieve document scores.
                                           If None, uses precomputed results in `queries_df`.
        file_name (str): Name of the output file where the run data will be saved.

    Behavior:
        - For each query, retrieves the top-k results using `retrieve_query_results` if `inv_index` is provided.
        - Formats the results in TREC run file format:
            `<query_id> Q0 <docno> <rank> <score> run.txt`
    """
    k = 10
    results = []

    if inv_index:
        for query in tqdm(queries_df.itertuples(index=False), desc='Creating run file', total=len(queries_df)):
            res, query_id = retrieve_query_results(query.query, inv_index, k), query.qid
            results.extend([f"{query_id} Q0 {r[1]} {rank} {r[0]} run.txt\n" for rank, r in enumerate(res, start=1)])
    else:
        for result in tqdm(queries_df.itertuples(index=False), desc='Creating run file', total=len(queries_df)):
            query_id, docno, rank, score = result.qid, result.docno, result.rank, result.score
            results.append(f"{query_id} Q0 {docno} {rank} {score} run.txt\n")

    with open(file_name, "w") as f:
        f.writelines(results)



## Generating TREC for benchmark
## Function: generate_trec_qrels_file  

### Purpose:
The `generate_trec_qrels_file` function generates a TREC-formatted qrels file from a dataset containing relevance judgments. This qrels file is used in IR evaluations to compare the system’s run file against human-annotated relevance judgments.

### Parameters:
- **dataset** (Dataset):  
  An object representing the dataset that contains relevance judgments. This object must provide a method `get_qrels("test-2020")` that returns a DataFrame with the relevance judgments.
  
- **file_name** (str):  
  The name (and optionally the path) of the output file where the qrels data will be saved.

### Behavior and Internal Workflow:
1. **Retrieve Relevance Judgments**:  
   - The function calls `dataset.get_qrels("test-2020")` to obtain a DataFrame (`qrels_df`) containing the relevance judgments.  
   - The DataFrame should have the following columns:
     - `qid`: The query identifier.
     - `docno`: The document identifier.
     - `label`: The relevance label (e.g., 0 or 1, or a graded relevance score).

2. **Formatting Qrels**:  
   - For each row in the relevance judgments DataFrame, the function formats a line according to the TREC qrels format:  
     `<query_id> 0 <docno> <label>`
   - Each formatted line is appended to a list named `qrels`.

3. **File Output**:  
   - The list of formatted qrels strings is written to the specified file (`file_name`) using standard file I/O operations.

### Returns:
- **None**:  
  The function writes the formatted qrels data directly to the specified file. It does not return any value, but its output is the generated qrels file in TREC format.

---

### Summary:
- **generate_trec_run_file**: Processes query to retrieve and rank documents using an inverted index, and writes the results in TREC run file format.
- **generate_trec_qrels_file**: Retrieves and formats relevance judgments from a dataset, writing them in TREC qrels file format.

Both functions serve critical roles in preparing the necessary files for IR evaluation, ensuring compatibility with standard evaluation protocols.


In [16]:
def generate_trec_qrels_file(dataset, file_name):
    """
    Generates a TREC qrels file from a dataset.

    Parameters:
        dataset (Dataset): Dataset object with a method `get_qrels("test-2020")` returning relevance judgments.
        file_name (str): Name of the output file where the qrels data will be saved.

    Behavior:
        - Retrieves relevance judgments as a DataFrame with columns 'qid', 'docno', and 'label'.
        - Formats the judgments in TREC qrels file format:
            `<query_id> 0 <docno> <label>`
    """
    qrels_df = dataset.get_qrels("test-2020")
    qrels = []

    for qrel in tqdm(qrels_df.itertuples(index=False), desc='Creating qrels file', total=len(qrels_df)):
        query_id, docno, label = qrel.qid, qrel.docno, qrel.label
        qrels.append(f"{query_id} 0 {docno} {label}\n")

    with open(file_name, "w") as f:
        f.writelines(qrels)

#Evaluation

## Calculating Evaluation Metrices

##  Function: run_trec_evaluation

### Purpose:
The `run_trec_evaluation` function orchestrates a full TREC evaluation process to measure the effectiveness of an Information Retrieval (IR) system. It generates a run file using a provided generator function, loads both the run file and the corresponding relevance judgments (qrels), and then computes a set of evaluation metrics. These metrics include **Precision at 5 (P@5)**, **nDCG at 10 (nDCG@10)**, **Average Precision (AP)**, and their graded variants (like AP(rel=2) and P(rel=2)@5) because they each capture different aspects of retrieval effectiveness. Since these are standard evaluation measures in IR research, they also allow the performance of my system to be compared with other systems in a consistent and reproducible manner.

### Parameters:
- **queries_df** (DataFrame):  
  A pandas DataFrame containing the queries. It typically includes at least:
  - `query`: The text of the query.
  - `qid`: The unique query identifier.

- **inv_index** (InvertedIndex):  
  An instance of the `InvertedIndex` class used to retrieve relevant documents for each query.  
  - This object provides the necessary posting lists and scoring functions for the retrieval process.

- **text_file** (str):  
  The file name (and optionally path) where the run file (i.e., the retrieval results) will be saved.  
  - The run file follows the TREC format and is used as input for evaluation.

- **run_file_generator** (function):  
  A function responsible for generating the run file. It is expected to accept the following parameters:
  - `queries_df`, `inv_index`, `text_file`, and any additional arguments provided via `*args`.
  - This function encapsulates the retrieval logic (e.g., calling `retrieve_query_results` or `daat`) and writes the formatted output to `text_file`.

- **\*args**:  
  Additional arguments that are forwarded to the `run_file_generator` function, allowing further customization of the run file generation process.

### Internal Workflow / Attributes:
  
1. **Loading Evaluation Files:**
   - The function then loads the relevance judgments (qrels) from a pre-specified file (`Files/qrels.txt`) using `ir_measures.read_trec_qrels`.
   - It also loads the run file generated earlier using `ir_measures.read_trec_run`.

2. **Aggregate Evaluation:**
   - Aggregate metrics are calculated using `ir_measures.calc_aggregate`, which computes measures such as:
     - Precision at 5 (P@5)
     - Precision for relevance 2 at 5 (P(rel=2)@5)
     - nDCG at 10 (nDCG@10)
     - Average Precision (AP)
     - Average Precision for relevance 2 (AP(rel=2))
   - These metrics are printed to provide an overview of the system's performance.

3. **Per-Query Evaluation:**
   - The function computes detailed per-query metrics using `ir_measures.iter_calc` for each measure.
   - The results are stored in dictionaries keyed by the query identifier (`query_id`), with each value representing the computed metric for that query:
     - `Pat5_list`: Precision at 5 per query.
     - `PRel2at5_list`: Precision for relevance 2 at 5 per query.
     - `nDCGat10_list`: nDCG at 10 per query.
     - `AP_list`: Average Precision per query.
     - `APRel2_list`: Average Precision for relevance 2 per query.

### Returns:
- **tuple**:  
  The function returns a tuple containing five dictionaries:
  1. **Pat5_list** (dict):  
     - Keys: Query IDs.  
     - Values: Precision at 5 for each query.
  2. **PRel2at5_list** (dict):  
     - Keys: Query IDs.  
     - Values: Precision at 5 for relevance level 2 for each query.
  3. **nDCGat10_list** (dict):  
     - Keys: Query IDs.  
     - Values: Normalized Discounted Cumulative Gain at 10 for each query.
  4. **AP_list** (dict):  
     - Keys: Query IDs.  
     - Values: Average Precision for each query.
  5. **APRel2_list** (dict):  
     - Keys: Query IDs.  
     - Values: Average Precision for relevance level 2 for each query.

### Summary:
The `run_trec_evaluation` function is a comprehensive evaluation routine that:
- Generates a TREC-formatted run file using a provided retrieval method.
- Loads the run file and corresponding qrels.
- Computes and prints aggregate evaluation metrics.
- Returns detailed per-query metrics for further analysis.
This function plays a critical role in quantifying the performance of an IR system in a standardized manner.

### Evaluation Metric Formulas

1. **Precision at K (P@K):**
   
   $$P@K = \frac{\text{Number of relevant documents in the top }K}{K}$$
   
   - *Interpretation:* This metric represents the fraction of the top K retrieved documents that are relevant to the query.

2. **Precision for Relevance 2 at K (P(rel=2)@K):**

   $$P(rel=2)@K = \frac{\text{Number of documents with relevance level 2 in the top }K}{K}$$
  
   - *Interpretation:* This measures the fraction of the top K documents that have a relevance label of 2.

3. **Normalized Discounted Cumulative Gain at K (nDCG@K):**
   
   $$\text{DCG@K} = \sum_{i=1}^{K} \frac{2^{rel_i} - 1}{\log_2(i+1)}$$
  
   $$\text{nDCG@K} = \frac{\text{DCG@K}}{\text{IDCG@K}}$$
   - *Interpretation:* DCG@K accumulates the gains from each relevant document, discounted by their rank. nDCG@K normalizes this value by the ideal DCG (IDCG@K), thus providing a measure of ranking quality that accounts for both relevance and position.

4. **Average Precision (AP):**
   $$AP = \frac{1}{R} \sum_{k=1}^{N} P(k) \times rel(k)$$
   - Where:
     - \( P(k) \) is the precision at rank \( k \),
     - \( rel(k) \) is an indicator function that equals 1 if the document at rank \( k \) is relevant, and 0 otherwise,
     - \( R \) is the total number of relevant documents for the query,
     - \( N \) is the total number of retrieved documents.
   - *Interpretation:* AP averages the precision values at the ranks where relevant documents occur, providing a single-number summary of the precision-recall curve.

5. **Average Precision for Relevance 2 (AP(rel=2)):**

   
 $$AP(rel=2) = \frac{1}{R_2} \sum_{k=1}^{N} P(k) \times I(rel(k)=2)$$
- Where:
- \( I(rel(k)=2) \) is an indicator function that equals 1 if the document at
 rank \( k \) has a relevance level of 2, and 0 otherwise, - \( R_2 \) is the
total number of documents with a relevance label of 2.
   - *Interpretation:* This metric calculates the average precision considering only those documents that have a relevance level of 2, offering insight into the performance for that specific relevance threshold.


In [17]:
from ir_measures import P, nDCG, AP
def run_trec_evaluation(queries_df, inv_index, text_file, run_file_generator, *args):
    """
    Executes a TREC evaluation for information retrieval performance.

    Parameters:
        queries_df (DataFrame): queries to be analyzed, typically containing queries and their IDs.
        inv_index (InvertedIndex): The inverted index used for retrieving documents.
        text_file (str): File name where the run file will be saved.
        run_file_generator (function): Function to generate the run file. Expected to accept:
            (queries_df, inv_index, text_file, *args).
        *args: Additional arguments to be passed to `run_file_generator`.

    Returns:
        tuple: Dictionaries containing evaluation metrics per query:
            - Pat5_list (dict): Precision at 5 (P@5) for each query.
            - PRel2at5_list (dict): Precision for relevance 2 at 5 (P(rel=2)@5).
            - nDCGat10_list (dict): Normalized Discounted Cumulative Gain at 10 (nDCG@10).
            - AP_list (dict): Average Precision (AP) for each query.
            - APRel2_list (dict): Average Precision for relevance 2 (AP(rel=2)).

    Prints:
        Aggregate evaluation metrics for the dataset using the specified measures.
    """

    # Load qrels and run files for evaluation
    qrels = list(ir_measures.read_trec_qrels('Files/qrels.txt'))
    run = list(ir_measures.read_trec_run(text_file))

    # Print aggregate evaluation metrics
    print(ir_measures.calc_aggregate([P@5, P(rel=2)@5, nDCG@10, AP, AP(rel=2)], qrels, run))

    # Compute per-query metrics
    Pat5_list = {m.query_id: m.value for m in ir_measures.iter_calc([P@5], qrels, run)}
    PRel2at5_list = {m.query_id: m.value for m in ir_measures.iter_calc([P(rel=2)@5], qrels, run)}
    nDCGat10_list = {m.query_id: m.value for m in ir_measures.iter_calc([nDCG@10], qrels, run)}
    AP_list = {m.query_id: m.value for m in ir_measures.iter_calc([AP], qrels, run)}
    APRel2_list = {m.query_id: m.value for m in ir_measures.iter_calc([AP(rel=2)], qrels, run)}

    return Pat5_list, PRel2at5_list, nDCGat10_list, AP_list, APRel2_list


In [18]:
import shutil

# Ensure NLTK stopwords are downloaded
nltk.download('stopwords', quiet=True)

# ===============================
# DEMO: SHOWCASING YOUR IR SYSTEM WITH BIGGER DOCUMENTS AND QUERIES
# ===============================

def demo_ir_system():
    # -------------------------------
    # 1. Create a Sample Dataset with Bigger Documents
    # -------------------------------
    docs = [
        {"docno": 1, "text": (
            "Machine learning is a rapidly evolving field of artificial intelligence. "
            "It involves the development of algorithms that can learn from data and make decisions "
            "with minimal human intervention. Recent advancements have led to significant breakthroughs "
            "in image recognition, speech processing, and autonomous driving. Researchers continue to explore "
            "new techniques to improve performance and scalability in real-world applications."
        )},
        {"docno": 2, "text": (
            "Deep neural networks represent a subset of machine learning algorithms that mimic the workings of the human brain. "
            "They are structured in layers that allow them to extract high-level features from raw data. This technology has revolutionized "
            "areas such as natural language processing, computer vision, and robotics by providing state-of-the-art performance in various tasks. "
            "Despite their power, these networks require significant computational resources and careful tuning."
        )},
        {"docno": 3, "text": (
            "Models in machine learning and neural networks are used for a wide range of applications, from classification and prediction to "
            "clustering and anomaly detection. With the explosion of big data, scalability has become a key factor. "
            "Recent developments include improved regularization techniques, optimization algorithms, and training methodologies that enable "
            "efficient processing of vast amounts of information, pushing the boundaries of what these models can achieve in both research and industry."
        )}
    ]
    df = pd.DataFrame(docs)

    # Preprocess the text for each document using your preprocess function
    # (This will tokenize, remove stopwords, and stem the text.)
    df["tokens"] = df["text"].apply(preprocess)

    print("=== Sample Documents and Their Tokens ===")
    print(df[["docno", "tokens"]])

    # -------------------------------------
    # 2. Build the Index Using Your Functions
    # -------------------------------------
    lexicon, inv, doc_index, direct_index, stats = build_index(df)

    print("\n=== Index Statistics ===")
    print(f"Number of documents: {stats['num_docs']}")
    print(f"Number of unique terms: {stats['num_terms']}")
    print(f"Total tokens in collection: {stats['num_tokens']}")

    print(inv)

    # Create an instance of your InvertedIndex class.
    # Choose BM25 parameters (here: k=1.2 and b=0.75).
    bm25_k = 1.2
    bm25_b = 0.75
    ir_index = InvertedIndex(lexicon, inv, doc_index, direct_index, stats, bm25_k, bm25_b)

    # -------------------------------------
    # 3. Run a Sample Query with Extended Text
    # -------------------------------------
    sample_query = (
        "machine learning applications in image recognition, speech processing, "
        "and autonomous systems with deep neural networks"
    )
    print(f"\n=== Running Query: '{sample_query}' ===")
    results = retrieve_query_results(sample_query, ir_index, heap_size=10)

    print("Retrieved Results (score, docid):")
    for score, docid in results:
        print(f"DocID: {docid}, Score: {score}")

    # --------------------------------------------------
    # 4. Generate a TREC-Formatted Run File for Sample Queries
    # --------------------------------------------------
    queries = [
        {"qid": 1, "query": (
            "machine learning applications in image recognition, speech processing, "
            "and autonomous systems with deep neural networks"
        )},
        {"qid": 2, "query": (
            "deep neural network architectures and natural language processing methods "
            "for scalable data analysis"
        )}
    ]
    queries_df = pd.DataFrame(queries)

    run_file_name = "sample_run.txt"
    generate_trec_run_file(queries_df, ir_index, run_file_name)

    print(f"\n=== Contents of the generated TREC run file '{run_file_name}' ===")
    with open(run_file_name, "r") as f:
        print(f.read())

    # --------------------------------------------------
    # 5. Create a Dummy QREL File for Evaluation
    # --------------------------------------------------
    # We simulate relevance judgments via a dummy dataset.
    class DummyDataset:
        def get_qrels(self, split):
            data = [
                {"qid": 1, "docno": 1, "label": 1},
                {"qid": 1, "docno": 3, "label": 1},
                {"qid": 2, "docno": 2, "label": 1},
                {"qid": 2, "docno": 3, "label": 0}
            ]
            return pd.DataFrame(data)

    dummy_dataset = DummyDataset()
    dummy_qrels_filename = "dummy_qrels.txt"
    generate_trec_qrels_file(dummy_dataset, dummy_qrels_filename)

    # The evaluation function expects the QRELs file in a folder "Files" as "Files/qrels.txt"
    if not os.path.exists("Files"):
        os.makedirs("Files")
    shutil.copy(dummy_qrels_filename, os.path.join("Files", "qrels.txt"))

    # --------------------------------------------------
    # 6. Run TREC Evaluation to Compute Metrics
    # --------------------------------------------------
    print("\n=== Running TREC Evaluation ===")
    metrics = run_trec_evaluation(queries_df, ir_index, run_file_name, generate_trec_run_file)

    # The function returns five dictionaries for per-query metrics.
    Pat5_list, PRel2at5_list, nDCGat10_list, AP_list, APRel2_list = metrics
    print("\n=== Per-Query Metrics ===")
    print("P@5:", Pat5_list)
    print("P(rel=2)@5:", PRel2at5_list)
    print("nDCG@10:", nDCGat10_list)
    print("AP:", AP_list)
    print("AP(rel=2):", APRel2_list)

# Execute the demonstration
demo_ir_system()


=== Sample Documents and Their Tokens ===
   docno                                             tokens
0      1  [machin, learn, rapid, evolv, field, artifici,...
1      2  [deep, neural, network, repres, subset, machin...
2      3  [model, machin, learn, neural, network, use, w...


Indexing: 100%|██████████| 3/3 [00:00<00:00, 2086.37it/s]



=== Index Statistics ===
Number of documents: 3
Number of unique terms: 104
Total tokens in collection: 131
{'docids': {0: [1, 2, 3], 1: [1, 2, 3], 2: [1], 3: [1], 4: [1], 5: [1], 6: [1], 7: [1], 8: [1, 3], 9: [1, 2, 3], 10: [1, 2, 3], 11: [1], 12: [1], 13: [1], 14: [1, 2], 15: [1], 16: [1, 3], 17: [1], 18: [1], 19: [1, 2], 20: [1], 21: [1], 22: [1], 23: [1], 24: [1, 2, 3], 25: [1], 26: [1], 27: [1, 3], 28: [1], 29: [1], 30: [1], 31: [1, 3], 32: [1, 3], 33: [1, 2], 34: [1, 3], 35: [1, 3], 36: [1], 37: [1], 38: [1, 3], 39: [2], 40: [2, 3], 41: [2, 3], 42: [2], 43: [2], 44: [2], 45: [2], 46: [2], 47: [2], 48: [2], 49: [2], 50: [2], 51: [2], 52: [2], 53: [2], 54: [2], 55: [2], 56: [2], 57: [2], 58: [2], 59: [2], 60: [2], 61: [2], 62: [2], 63: [2], 64: [2], 65: [2], 66: [2], 67: [2], 68: [2], 69: [2], 70: [2], 71: [2], 72: [2], 73: [2], 74: [3], 75: [3], 76: [3], 77: [3], 78: [3], 79: [3], 80: [3], 81: [3], 82: [3], 83: [3], 84: [3], 85: [3], 86: [3], 87: [3], 88: [3], 89: [3], 90: [3], 9

Creating run file: 100%|██████████| 2/2 [00:00<00:00, 1262.77it/s]



=== Contents of the generated TREC run file 'sample_run.txt' ===
1 Q0 1 1 2.2593919095374027 run.txt
1 Q0 2 2 0.9264948826648859 run.txt
1 Q0 3 3 0.541079074657075 run.txt
2 Q0 2 1 1.9129116737239835 run.txt
2 Q0 3 2 0.3607193831047167 run.txt



Creating qrels file: 100%|██████████| 4/4 [00:00<00:00, 35544.95it/s]


=== Running TREC Evaluation ===
{AP: 0.9166666666666666, AP(rel=2): 0.0, P@5: 0.30000000000000004, nDCG@10: 0.9598603945740938, P(rel=2)@5: 0.0}

=== Per-Query Metrics ===
P@5: {'1': 0.4, '2': 0.2}
P(rel=2)@5: {'1': 0.0, '2': 0.0}
nDCG@10: {'1': 0.9197207891481876, '2': 1.0}
AP: {'1': 0.8333333333333333, '2': 1.0}
AP(rel=2): {'1': 0.0, '2': 0.0}


## Conclusion

The experimental results obtained on a small-scale dataset demonstrate that the IR system performs all critical operations successfully—from document preprocessing and indexing to query retrieval, run file generation, and evaluation using standard IR metrics. Below is a summary of the outcomes:

### Preprocessing & Tokenization
- **Documents Processed:** Three sample documents were processed.
- **Tokens Generated:** A total of 131 tokens were produced, with 104 unique terms.
- **Effectiveness:** The preprocessing step (including lowercasing, stopword removal, and stemming) effectively reduced the documents to their essential components, preparing them for efficient indexing.

### Indexing
- **Index Structures:** The inverted index, lexicon, and other supporting structures were built correctly, ensuring efficient retrieval of documents.

### Query Processing & Ranking
- **Sample Query:**  
  *"machine learning applications in image recognition, speech processing, and autonomous systems with deep neural networks"*
- **Retrieval Results:**  
  - **DocID 1:** Score = 2.2594  
  - **DocID 2:** Score = 0.9265  
  - **DocID 3:** Score = 0.5411  
- **Analysis:** The ranking indicates that Document 1 is the most relevant based on the query, which aligns with the content of the document.

### Run File Generation
- **TREC Format:** A TREC-formatted run file was successfully generated for two sample queries.
- **Format Compliance:** The run file follows the format `<query_id> Q0 <docno> <rank> <score> run.txt`, making it compatible with standard IR evaluation tools.
- **File Contents:** The generated file correctly lists the results for each query in the expected format.

### Evaluation Metrics
- **Aggregate Metrics:**
  - **Precision at 5 (P@5):** 0.30  
    *(On average, 30% of the top 5 retrieved documents are relevant.)*
  - **Normalized Discounted Cumulative Gain at 10 (nDCG@10):** 0.96  
    *(Indicates nearly optimal ranking quality.)*
  - **Average Precision (AP):** 0.92  
    *(Reflects high overall retrieval precision.)*
  - **AP(rel=2) and P(rel=2)@5:** 0.0  
    *(No documents with a relevance label of 2 were present in this sample.)*
  
- **Per-Query Metrics:**
  - **P@5:**  
    - Query 1: 0.4  
    - Query 2: 0.2
  - **nDCG@10:**  
    - Query 1: 0.92  
    - Query 2: 1.0
  - **AP:**  
    - Query 1: 0.83  
    - Query 2: 1.0

These metrics validate the effectiveness of the retrieval system, particularly in terms of ranking quality (as indicated by the high nDCG@10 and AP values). The lower precision at 5, however, suggests that there is room for improvement in retrieving a higher proportion of relevant documents within the top ranks.

### Summary
In summary, the demonstration confirms that the IR system is fully functional:
- **Efficient Processing:** It efficiently preprocesses and indexes documents.
- **Accurate Retrieval:** It handles complex queries and produces meaningful ranked results.
- **Robust Evaluation:** It generates a TREC-formatted run file and computes standard IR metrics that provide insights into system performance.

Although this test was conducted on a small sample, it serves as a proof-of-concept for scaling the system to larger datasets (such as the intended 2GB collection) when adequate computational resources are available.


In [19]:
# !pip install swifter

In [20]:
!pip install dask distributed

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 16.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 35.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.3/43.3 kB 2.9 MB/s eta 0:00:00
  Attempting uninstall: dask
    Found existing installation: dask 2024.10.0
    Uninstalling dask-2024.10.0:
      Successfully uninstalled dask-2024.10.0


In [21]:
# import dask.dataframe as dd
# from dask.distributed import Client
# # ----------------------------------------------------
# # 3. Set up Dask and process the data
# # ----------------------------------------------------
# # Optionally, you can start a Dask distributed client to better monitor performance.
# client = Client()  # This starts a local Dask cluster

# # Define the path to your TSV file.
# tsv_file = '/content/drive/MyDrive/msmarco_dataset/collection/collection.tsv'

# # Load the TSV file into a Dask DataFrame.
# # Adjust parameters if your file has a header. Here, we assume it does not.
# ddf = dd.read_csv(tsv_file,
#                   sep='\t',
#                   header=None,
#                   names=['docno', 'text'],
#                   encoding='utf-8')

# # Apply the preprocess function to the "text" column.
# # The meta parameter tells Dask about the output type.
# ddf['tokens'] = ddf['text'].apply(preprocess, meta=('tokens', 'object'))

# # Trigger computation. This will run the tasks in parallel.
# result_df = ddf.compute()

# # Optionally, save the result to disk.
# result_df.to_csv('processed_dataset.csv', index=False)

# print("Processing complete!")

In [22]:
# import swifter
# import multiprocessing as mp
# # Set this flag to True to use multiprocessing; set to False to try swifter instead.
# USE_MULTIPROCESSING = False

# def process_chunk(chunk):
#     """
#     Processes a DataFrame chunk by applying the preprocess function to its "text" column.
#     """
#     if USE_MULTIPROCESSING:
#         # Convert the text column to a list for faster mapping in multiprocessing
#         texts = chunk["text"].tolist()
#         with mp.Pool(processes=mp.cpu_count()) as pool:
#             tokens = pool.map(preprocess, texts)
#         chunk["tokens"] = tokens
#     else:
#         # Alternative: Use swifter for a potentially parallelized apply
#         # Make sure swifter is installed and imported if you want to use this branch.
#         chunk["tokens"] = chunk["text"].swifter.apply(preprocess)
#     return chunk

# # -------------------------------
# # 3. Main processing function
# # -------------------------------

# def main():
#     # Path to your TSV file (adjust the path as needed)
#     tsv_file = '/content/drive/MyDrive/msmarco_dataset/collection/collection.tsv'

#     # Set the chunk size (adjust based on your memory and file size)
#     chunk_size = 100000
#     processed_chunks = []

#     # Read and process the file in chunks using pandas
#     for chunk in pd.read_csv(tsv_file,
#                              sep='\t',
#                              header=None,
#                              names=['docno', 'text'],
#                              chunksize=chunk_size,
#                              encoding='utf-8'):
#         processed_chunk = process_chunk(chunk)
#         processed_chunks.append(processed_chunk)

#         # Clean up to free memory
#         gc.collect()

#     # Concatenate all processed chunks into a single DataFrame
#     dataset = pd.concat(processed_chunks, ignore_index=True)
#     print("Processing complete!")

#     # Optionally, save the processed dataset to disk
#     # dataset.to_csv('processed_dataset.csv', index=False)

# if __name__ == '__main__':
#     main()


In [23]:
# tsv_file = '/content/drive/MyDrive/msmarco_dataset/collection/collection.tsv'
# dataset = pl.read_csv(tsv_file, separator='\t', has_header=False, new_columns=['docno', 'text'], encoding="utf-8").to_pandas()


In [24]:
# dataset["tokens"] = dataset["text"].apply(preprocess)
# gc.collect()

In [25]:
#  print("Building index...")
#  lex, inv, doc, direct, stats = build_index(dataset)
#  print(stats)

In [26]:
# inv_index = InvertedIndex(lex, inv, doc, direct, stats, 1.2, 0.7)